src/backend/executor/execTuples.c

```diff
@@ -751,6 +751,53 @@ ExecFetchSlotHeapTuple(TupleTableSlot *slot)
 	return htup;
 }
 
+HeapTuple
+ExecMaterializeSlotHeapTuple(TupleTableSlot *slot)
+{
+	uint32 tuplen;
+	HeapTuple htup;
+
+	/*
+	 * sanity checks
+	 */
+	Assert(!TupIsNull(slot));
+
+	/*
+	 * If we have a regular physical tuple, and it's locally palloc'd, we have
+	 * nothing to do.
+	 */
+	if (slot->PRIVATE_tts_heaptuple && slot->PRIVATE_tts_htup_buf)
+		return slot->PRIVATE_tts_heaptuple;
+
+	slot_getallattrs(slot);
+
+	Assert(TupHasVirtualTuple(slot));
+	Assert(slot->PRIVATE_tts_nvalid == slot->tts_tupleDescriptor->natts);
+
+	/* Get the length of the buffer we need to allocate. */
+	tuplen = 0;
+	htup = heaptuple_form_to(slot->tts_tupleDescriptor, slot_get_values(slot),
```

Comment:
From the code, memtuple refer to the internal buffer of appendonly scanner if we call appendonly_getnext for append table. Should we also copy for memtuple because modification on memtuple also change buffer state of appendonly scanner?

Reply:
Is this a question for the current patch, or for a future patch that adds materialization support for memtuples? I didn't think this code path touched memtuples.